<a href="https://colab.research.google.com/github/BigFoots625/IntroductionMachineLearningwithPython/blob/main/Chapter1_Introduction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **# Chapter 1: Introduction**
In this chapter, we introduce the core concepts of machine learning, set up our Python environment, and build our first model using the famous Iris dataset. The goal is to predict the species of an iris flower based on the physical measurements of its petals and sepals.

In [1]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split

# Load the dataset
iris_dataset = load_iris()
print("Keys of iris_dataset:\n", iris_dataset.keys())
print("Target names:", iris_dataset['target_names'])
print("Feature names:\n", iris_dataset['feature_names'])

Keys of iris_dataset:
 dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])
Target names: ['setosa' 'versicolor' 'virginica']
Feature names:
 ['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


# **Theoretical Explanation: Why Split the Data?**
Before building a machine learning model, we must split our data into a **training set** and a **test set** (often a 75/25 split).

If we train and evaluate the model on the exact same data, the model might just memorize the answers (a problem known as *overfitting*). By holding out a test set, we can evaluate how well our model **generalizes** to new, unseen data.

In [2]:
X_train, X_test, y_train, y_test = train_test_split(
    iris_dataset['data'], iris_dataset['target'], random_state=0)

print("X_train shape:", X_train.shape)
print("y_train shape:", y_train.shape)
print("X_test shape:", X_test.shape)
print("y_test shape:", y_test.shape)

X_train shape: (112, 4)
y_train shape: (112,)
X_test shape: (38, 4)
y_test shape: (38,)


### **Theoretical Explanation: k-Nearest Neighbors (k-NN)**
The k-NN algorithm is one of the simplest classification models. Building the model consists only of storing the training dataset.

To make a prediction for a new data point, the algorithm finds the point in the training set that is closest to the new point. Then, it assigns the label of this training point to the new data point. The "k" in k-NN signifies that instead of using only the closest neighbor, we can consider any fixed number $k$ of neighbors and make a prediction based on the majority class among them.

Distance is typically calculated using the Euclidean distance formula:
$$d(p, q) = \sqrt{\sum_{i=1}^{n} (q_i - p_i)^2}$$

In [3]:
from sklearn.neighbors import KNeighborsClassifier

# Instantiate the model with k=1
knn = KNeighborsClassifier(n_neighbors=1)

# Train the model
knn.fit(X_train, y_train)

# Make a prediction on a new, hypothetical flower
X_new = np.array([[5, 2.9, 1, 0.2]])
prediction = knn.predict(X_new)
print("Prediction:", prediction)
print("Predicted target name:", iris_dataset['target_names'][prediction])

# Evaluate the model on the test set
y_pred = knn.predict(X_test)
print("Test set score: {:.2f}".format(knn.score(X_test, y_test)))

Prediction: [0]
Predicted target name: ['setosa']
Test set score: 0.97


### **Chapter 1 Summary**
* **Objective:** We successfully built a machine learning model to classify iris species.
* **Key Concepts:** We learned the importance of splitting data into training and testing sets to ensure our model can generalize to new data.
* **Algorithm Used:** We utilized the k-Nearest Neighbors (k-NN) classifier, an instance-based learning algorithm that classifies new data points based on their proximity to existing training data.
* **Result:** Our model achieved an accuracy of 97% on the test set, giving us confidence in its predictive capabilities.